### ISSUE TITLE
Parallelize the rsync run using multiple threads and/or connections (old Bugzilla bug 5124)

### ISSUE THREAD

**Issue by lockywolf** [WayneD/rsync#131]
I am not sure it is appropriate to manually refile bugs from the old Bugzilla to the new issue tracker at GitHub, but there is an issue that is largely a show-stopper for me and for many other users, so if possible I'd like to "modernise" the issue but moving it here.

https://bugzilla.samba.org/show_bug.cgi?id=5124

The idea is to use multiple TCP connections to download a single piece of data by splitting it into N pieces and downloading them in parallel. 

This used to be less useful when the Internet was young, and multiple connections would only help to save on TCP connection establishing time with many tiny files. However, nowadays several incredibly huge ISPs are clamping non-DPI-able connections to 10k/s or so. Rsync, being done through SSH, obviously belongs to  "non-DPI-able" connections and therefore becomes almost useless unless it can work in parallel and obtain an Nthreads*10k/s speed. 

I would, therefore, ask for this feature to be implemented. The algorithm already used by Aria2 (and its --split) option is likely to be a good candidate.

**Comment by lockywolf** [comment]
In fact, aria2 can even do downloading  through ssh, as documented here: https://github.com/aria2/aria2/issues/453

**Comment by hardhub** [comment]
> This would be extremely useful, as for high-speed links (10/40/100Gbit)

@dsg22 That will be extremely useful for slow links as well, if you uploading through a lot of hopes and ISPs which shapes traffic.

**Comment by sheepdestroyer** [comment]
This seems like a huge oversight, and despite searching, I haven't seen any reasoning for why this feature hasn't been implemented. What would be the rationale for a "single stream by design" ?

**Comment by WayneD** [comment]
It requires a complete rewrite of the utility using a totally new protocol stream. Something that I've been considering, but is not simple nor going to happen any time soon.

**Comment by bahbarnett** [comment]
I would suggest that this could be done without protocol changes.

- an initial connection with a single local->remote instance, which scans the destination filesystem (I believe rsync already does this)
- once complete, the parent rsync could spawn $x sub-rsync processes, which all simply run, but take a filename (or file list) from the parent (eg, socket, or stdin, or some such)

The above could be done in two stages.  

1) The first would only allow non-destructive operations, that is, simply "move data".

2) After the first run, rsync could drop all children, and re-scan the filesystem according to any destructive operations.  This will ensure that --delete, and other operations, are running once a 100% verification of how the remote state is left, is done.

This is the manual process many of us are following.

It may seem hacky, but really, I think any parallel operation via rsync will require spawning sub-rsync + ssh sessions regardless.  So having rsync simply spawn itself with alternate command line args, would work well here.

Even disabling some aspects of rsync for this mode would be well understood, I think.  That is, "you cannot use destructive operations when doing parallel transfers", or some such.

So users could run in parallel mode, shove the data across, then re-run in 'destructive operations' mode, eg --delete and other flags of this nature.

**Comment by WayneD** [comment]
That is something that someone could write as a wrapping script (and indeed, is something that I have implemented before where my script copied directory trees in order of largest to smallest, keeping a certain number of parallel rsyncs active at the same time. This is not something that rsync itself should do, though, as the extra scan of the source files would be a bad idea -- the user's script should use its own cached knowledge of what to copy in what order and could even update the sizes based on the output of the rsync copies (which is a reasonable starting point for the next copy). An actual parallel rsync would farm out individual files to a parallel set of rsyncs, as that doesn't require any extra scanning or knowledge but does require a new protocol.

**Comment by michaelmess-de** [comment]
When spawning additional ssh threads, this should be done for large files only as spawning processes itself is costly.
But the biggest drawback is: When the user logs in using password login instead of key login, the user has to enter his password for each spawned process again and again, which could be annoying. We would then also have to handle the case when the password was not correct and login for that thread failed, which makes it even more complicated.

I think rsync is not the right place to do that and ssh should be improved for its performance instead.

Perhaps it would be beneficial to share the work of encrypting the data between multiple threads - the decision to do that should be done locally by the ssh client/server - on a low performance cpu with multiple cores the work could then be distributed while on a high performance cpu only one core might be sufficient to handle the work alone so that no distribution is necessary.
And with ssh, not only rsync gets a benefit from it, but all applications that use remote connections or port forwarding via ssh would benefit from that improvement.

**Comment by michaelmess-de** [comment]
An optimisation for ssh for high throughput already exists and can be found here:
https://github.com/rapier1/hpn-ssh

**Comment by michaelmess-de** [comment]
I just tested the performance of hpnssh

```
# Iperf direct
michael@thinkpad:~$ iperf -c localhost -p 5001  
------------------------------------------------------------
Client connecting to localhost, TCP port 5001
TCP window size: 16.0 KByte (default)
------------------------------------------------------------
[  1] local 127.0.0.1 port 55746 connected with 127.0.0.1 port 5001 (icwnd/mss/irtt=319/32741/34)
[ ID] Interval       Transfer     Bandwidth
[  1] 0.0000-10.0083 sec  17.9 GBytes  15.4 Gbits/sec
michael@thinkpad:~$ 

# Iperf with ssh 
ssh -L 5002:localhost:5001 localhost

michael@thinkpad:~$ iperf -c localhost -p 5002  
------------------------------------------------------------
Client connecting to localhost, TCP port 5002
TCP window size: 2.50 MByte (default)
------------------------------------------------------------
[  1] local 127.0.0.1 port 41066 connected with 127.0.0.1 port 5002 (icwnd/mss/irtt=319/32741/29)
[ ID] Interval       Transfer     Bandwidth
[  1] 0.0000-10.0322 sec  2.86 GBytes  2.45 Gbits/sec
michael@thinkpad:~$ 

# Iperf with hpnssh-client (ssh-server)
michael@thinkpad:~$ iperf -c localhost -p 5003  
------------------------------------------------------------
Client connecting to localhost, TCP port 5003
TCP window size: 16.0 KByte (default)
------------------------------------------------------------
[  1] local 127.0.0.1 port 37540 connected with 127.0.0.1 port 5003 (icwnd/mss/irtt=319/32741/53)
[ ID] Interval       Transfer     Bandwidth
[  1] 0.0000-10.0290 sec  2.77 GBytes  2.37 Gbits/sec
michael@thinkpad:~$

# IPerf with hpnssh-client and server
michael@thinkpad:~$ iperf -c localhost -p 5003  
------------------------------------------------------------
Client connecting to localhost, TCP port 5003
TCP window size: 2.50 MByte (default)
------------------------------------------------------------
[  1] local 127.0.0.1 port 54132 connected with 127.0.0.1 port 5003 (icwnd/mss/irtt=319/32741/36)
[ ID] Interval       Transfer     Bandwidth
[  1] 0.0000-10.0308 sec  3.16 GBytes  2.71 Gbits/sec
michael@thinkpad:~$ 
```
Maybe the numbers are not accurate, because there are other things (like the Web browser) running on the host, too. 
For accurate numbers I would have to repeat this with no other things running and do multiple test runs to get an average.

But the numbers show that the performance gains are lower than I have expected.
